### PR TITLE
Import rework

### LINKED ISSUE #2951: Importing shared decks breaks local decks (if names are the same)
Originally reported on Google Code with ID 2054
```
What steps will reproduce the problem?

- Import a shared deck which has the same name as a local deck.


What is the expected output? What do you see instead?

- The shared and the local decks get mixed up somehow. The app flags the deck as imported
shared deck but it still shows the local decks cards. It might simply add the shared
decks cards, I'm not sure about that.
Expected behaviour would be to show a warning that the deck name is already in use
and ask for a different deck name to use for the import.


Does it happen again every time you repeat the steps above? Or did it
happen only one time?

- Happens all the time. I tried it with the shared deck "BWI". My local BWI deck has
lots of sub decks but that's probably not relevant.


What version of AnkiDroid are you using? (Decks list > menu > About > Look
at the title)

- 2.1.1


On what version of Android? (Home screen > menu > About phone > Android
version)

- 4.4.2


If it is a crash or "Force close" and you can reproduce it, the following
would help immensely: 1) Install the "SendLog" app, 2) Reproduce the crash,
3) Immediately after, launch SendLog, 4) Attach the resulting file to this
report. That will make the bug much easier to fix.


```

Reported by `gudi852` on 2014-03-31 15:47:03

### LINKED ISSUE #3248: Loading shared deck: takes a long time even after all GUI marks get checked
Originally reported on Google Code with ID 2352
```
What steps will reproduce the problem?
1. Download a shared deck via Firefox
2. Open the download (pick AnkiDroid if asked)
3. Watch the import dialog in AnkiDroid
4. See that all four items in that dialog are checked or at 100%

What is the expected output? What do you see instead?

When the fourth line got its check mark, i assumed that the import was finished, but
it still took quite a while.

I then pressed the home button and went back to AnkiDroid. The import dialog was gone,
but apparently the import wasn't finished. 

Then i touched "Sync" and after a while got "Sync finished" an an empty deck overview.

Does it happen again every time you repeat the steps above? Or did it
happen only one time?

The empty deck overview happened only once.

What version of AnkiDroid are you using? (Decks list > menu > About > Look
at the title)
On what version of Android? (Home screen > menu > About phone > Android
version)

Android 4.1
AnkiDroid 2.3beta44

If it is a crash or "Force close" and you can reproduce it, the following
would help immensely: 1) Install the "SendLog" app, 2) Reproduce the crash,
3) Immediately after, launch SendLog, 4) Attach the resulting file to this
report. That will make the bug much easier to fix.


Please provide any additional information below.

```

Reported by `ospalh` on 2014-10-27 11:12:07

<hr>
 * *Attachment: import_check.png<br>*
 * *Attachment: Screenshot_2014-10-27-11-01-13.png<br>*

### LINKED ISSUE #3335: Existing notes are not updated on import
Originally reported on Google Code with ID 2439
```
What steps will reproduce the problem?
1. having ankidroid update the changes onto the phone
2.
3.

What is the expected output? What do you see instead?
i expect to see changes updated, instead no changes occur

Does it happen again every time you repeat the steps above? Or did it
happen only one time?
everytime


What version of AnkiDroid are you using? (Decks list > menu > About > Look
at the title)
On what version of Android? (Home screen > menu > About phone > Android
version)
v2.3.2

If it is a crash or "Force close" and you can reproduce it, the following
would help immensely: 1) Install the "SendLog" app, 2) Reproduce the crash,
3) Immediately after, launch SendLog, 4) Attach the resulting file to this
report. That will make the bug much easier to fix.


Please provide any additional information below.

```

Reported by `kawaharajamie` on 2014-12-14 04:25:01

### LINKED ISSUE #3539: X cards imported dialog should use plurals
Originally reported on Google Code with ID 2667
```
Note: If this is your first time submitting an issue, please read the
following webpage:

https://ankidroid.org/docs/help.html


What steps will reproduce the problem?
1. Import a deck with just a single card.¹
2. Look at the dialog after the import


What is the expected output? What do you see instead?

The text should use a singular form.
Instead you get the one-case-fits-all “import_log_success” string “Imported 1 cards”

Does it happen again every time you repeat the steps above? Or did it
happen only one time?

Every time

Paste the "debug info" from the AnkiDroid "About" dialog below (see the
above help webpage for detailed instructions).

I’ll skip this. The text has been this way forever.

Please provide any additional information below.

¹ This problem is of course not visible when you use an UI language that doesn’t use
plurals, like Japanese, but probably *is* visible for other card counts in e.g. Eastern
European languages with more complicated rules.
```

Reported by `ospalh` on 2015-06-29 08:30:51

### LINKED ISSUE #3832: Crash when importing due to ending an unopen transaction
It looks like we're closing the collection at one point inside the importing routine which is interfering with the rollback mechanism we have in DeckTask. Maybe we need to move the rollback / transaction code into the importing routine itself.

android.database.sqlite.SQLiteException: cannot rollback - no transaction is active (code 1)
at android.database.sqlite.SQLiteConnection.nativeExecute(Native Method)
at android.database.sqlite.SQLiteConnection.execute(SQLiteConnection.java:552)
at android.database.sqlite.SQLiteSession.endTransactionUnchecked(SQLiteSession.java:439)
at android.database.sqlite.SQLiteSession.endTransaction(SQLiteSession.java:401)
at android.database.sqlite.SQLiteDatabase.endTransaction(SQLiteDatabase.java:522)
at com.ichi2.async.DeckTask.doInBackgroundImportAdd(DeckTask.java:801)
at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:280)
at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:66)

### PR DESCRIPTION
**Pull request by hssm**
This is a rework of the importer code. The structure of the importers was rewritten to match the desktop version and a bunch of things were fixed. This should also make it somewhat easier to port other importers later on.

Fixes #3832 - Transactions are entirely within the importer now with proper try/finally blocks.
Fixes #3539 - I resorted to the "Cards imported: X" style we're using everywhere else. Also made some changes to a few other strings.
Fixes #3335 - All the features in the original are ported over, so updates and dupe handling are all working now.
Fixed #2951 - It's an intended feature, but our existing implementation of it was broken. It's fixed now (as above).
Fixes #3248 - The work being done after the check marks was importing static media and vacuum/analyze which can take a while. I reworked progress steps and the completion chunks so it should continue to progress until it's actually 100%.

I've kept the media optimizations in place. I feel like the 1KB dupe check might fail in some cases but it's probably extremely unlikely in practice.

Since the importer makes use of localized strings, we need access to a Context. I've made it so we include a Context object when creating a Collection so we can have access to it inside libanki. I avoided this last time because of its ties with the importer, but since I'm going over it now it seemed like the right time to do it.

Doing this clashed with the `reopenCollection()` method in `CollectionHelper`. The javadoc states that its purpose is for when we want to reopen the collection when we don't have a Context which goes against what we are trying to do now. After some investigation I've understood what the method was trying to do and ended up removing it along with some TODOs in the sync code. Basically the full syncer closes the collection and then needs to reopen it to progress with the media sync (since the media DB is opened and closed through the main DB). There's a method in Collection for doing that already. This portion of the sync code is now more similar to the desktop version.

### CODE REVIEW

**Comment by hssm** [comment]
I've updated this commit and the original PR.

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.java [hunk 1]
```diff
@@ -455,14 +451,9 @@ private Payload doInBackgroundSync(Payload data) {
             }
             return data;
         } finally {
-            if (col != null) {
-                // Close collection to roll back any sync failures
```
Comment (hssm) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.java]: This comment is misleading (and trailed off until a few days ago). The normal syncer uses transactions so rollbacks would happen in there, and the full syncer uses a temporary file so there is no rolling back to do.

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.java [hunk 2]
```diff
@@ -455,14 +451,9 @@ private Payload doInBackgroundSync(Payload data) {
...
-                Timber.d("doInBackgroundSync -- closing collection on outer finally statement");
-                col.close(false);
-            }
+            // don't bump mod time unless we explicitly save
+            col.close(false);
```
Comment (hssm) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.java]: And this line becomes the equivalent to [this](https://github.com/dae/anki/blob/master/aqt/sync.py#L331) in the original.

**Comment by timrae** [comment]
Is it possible to store the test apkg files somewhere else? If they go in the assets folder then they'll be compiled into the apk binary.

**Comment by hssm** [comment]
That's a good point. I *think* the test project can have its own assets folder. I'll see if that works.

File: AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java [hunk 3]
```diff
@@ -83,7 +83,7 @@ public static CollectionHelper getInstance() {
      * @param path path to the AnkiDroid file
      * @return instance of the Collection
      */
-    private synchronized Collection getCol(String path) {
+    private synchronized Collection getCol(String path, Context context) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java]: you should probably add the new argument to the javadoc ;)

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java]: In fact, I don't think we require two methods here anymore....

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java [hunk 4]
```diff
@@ -36,286 +36,243 @@
...
-            mProgress.publishProgress(new DeckTask.TaskData(mResources.getString(R.string.import_add_progress,
-                    (unpacking ? CHECKMARK : "-"), notesDone, cardsDone, (cleanup ? CHECKMARK : "-"))));
-        }
+    /**
+     * Since we can't use a tuple as a key in Java, we resort to indexing twice with nested maps.
+     * Python: (guid, ord) -> cid
+     * Java: guid -> ord -> cid
+     */
+    private HashMap<String, HashMap<Integer, Long>> mCards;
+    private HashMap<Long, Long> mDecks;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: It probably doesn't make much difference, but in terms of [best practices](http://stackoverflow.com/questions/19859750/is-it-a-good-common-practice-to-use-abstract-classes-in-polymorphism-in-java) can we declare these as the more generic `Map` objects?

File: AnkiDroid/src/main/java/com/ichi2/async/Connection.java (hunk 2 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/async/Connection.java]: It sure would be nice to someday have a more clear partition between the libanki sync code and the Android async code

**Comment by timrae** [comment]
I only had time to do a cursory review, but it looks great!

File: AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java (hunk 3 above)
Comment (hssm) [inline AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java]: Yep, I'll merge them into one.

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java (hunk 4 above)
Comment (hssm) [inline AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java]: Yeah it's a good habit to get into. I'm normally pretty careful with these if they're publicly accessible.

**Comment by hssm** [comment]
The commit has been updated after the comments. The assets are moved into the test project now, and I realised we could now move the hook initialization into Collection now since we *do* have a Context.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java.orig [hunk 5]
```diff
@@ -0,0 +1,445 @@
+
```
Comment (hssm) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java.orig]: Wait this file shouldn't be here.

**Pull request merged**